### DISCUSSION TITLE
Sass-loader can't resolve background-image url

### DISCUSSION THREAD

**Discussion by JinHe-PerkinElmer** [webpack/webpack#15073] [Q&A]
I' trying to upgrade webpack from v4 to v5. The dependencies have been upgraded according to the official upgrade document, but all the background image url inside scss files can't be resolved. Error like this:
```
ERROR in ./src/modules/userManagement/components/editUserModal.scss 
(./node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[0].oneOf[5].use[1]!./node_modules/postcss-loader/dist/cjs.js??
ruleSet[1].rules[0].oneOf[5].use[2]!./node_modules/resolve-url-loader/index.js??
ruleSet[1].rules[0].oneOf[5].use[3]!./node_modules/sass-loader/dist/cjs.js??
ruleSet[1].rules[0].oneOf[5].use[4]!./src/modules/userManagement/components/editUserModal.scss)

Module not found: Error: Can't resolve '/images/btn_icon_filter_selected.png' in 'C:\Users\HeJ17302\Documents\GitHub\OneSource-Portal\NewPortal\SPA\src\modules\userManagement\components'
 @ ./src/modules/userManagement/components/editUserModal.scss
```
webpack.config.js like this: 
```
module.exports = {
  module: {
    rules: [
      {
        test: /\.s[ac]ss$/i,
        use: [
          "style-loader",
          "css-loader",
          "postcss-loader",
          "resolve-url-loader",
          "sass-loader",
        ],
      },
    ],
  },
};
```
And I have tried 'resolve-url-loader', it can't help. Anyone can help?

**Comment by alexander-akait** [comment]
I think it is not related to sass-loader, `css-loader` now handles `/path/to/image.png` URLs, you can ignore them using https://github.com/webpack-contrib/css-loader#object or provide path using the `resolver.alias` option

**Reply by JinHe-PerkinElmer** [reply]
Thanks for your reply. But I think resolver.alias is not suitable for me(Too many images). In fact I use @minxin in my global.scss file:
```
@mixin getBGImgPath($imgName, $important:false) {
  @if $important {
    background: url("#{$baseURL}#{$imgName}") no-repeat!important;
  } @else {
    background: url("#{$baseURL}#{$imgName}") no-repeat;
  }
}
```
And in many other scss file to use it after @import '../../../global.scss'
`@include getBGImgPath('btn_icon_filter_default.png');`
All background images can be resolved normally before the upgrade. So I guess this a problem with sass-loader or css-loader.

**Reply by alexander-akait** [reply]
No problems with css-loader or sass-loader, we handle `/path/to/image.png` URLs, so you need to fix path to images or ignore them, also you can set valid `baseURL` (you need specify path where you store images)

**Reply by JinHe-PerkinElmer** [reply]
@alexander-akait Thanks for your reply very much. I have set my background-image baseURL like '/images/'. Now the problem is maybe the css-loader resolve .png file with wrong path. Here is the full error msgs: 
```
[webpack-dev-server] ERROR in ./src/index.build.test.scss (./node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[0].oneOf[5].use[1]!./node_modules/sass-loader/dist/cjs.js??ruleSet[1].rules[0].oneOf[5].use[2]!./src/index.build.test.scss) 7:36-98
Module not found: Error: Can't resolve '/images/icon_L_M_arrow_default.png' in 'C:\Users\HeJ17302\Documents\GitHub\OneSource-Portal\NewPortal\SPA\src'
```

Attention please, my project is based on create-react-app. So my webpack entry is **SPA/src/index.jsx**, and image files is under **SPA/public**

**Reply by alexander-akait** [reply]
It means you still have wrong path on images, if you provide simple example I will provide code how to solve it

**Reply by JinHe-PerkinElmer** [reply]
Yes, of course. Here is a simple demo: https://codesandbox.io/s/fast-mountain-649pv?file=/src/index.build.test.jsx
I think it is necessary to explain:
1. Project is created by create-react-app template, then we ejected it for customization;
2. We start webpack compiling and webpack-dev-server through node JS API.(So the configuration file looks more complicated).
3. The demo is very simple, just want to use background-image in scss file. Of course, you can only see the code part in this online tool, not the page effect. Please let me know if you have any questions. Thanks for your help.

**Reply by alexander-akait** [reply]
Hm, why you put source image in `public/images`? Yep, you can do it, but I recommend to put images inside `src` (i.e. `src/images`), why? Because it is source image, webpack generates assets and copy them in dist/build/public, `public` directory most commonly used for fully static assets, i.e. `favicon.ico`/`robots.txt`/etc, also in future you can want to generate multiple images for example, webp/avif there they supported, so maybe sense to keep original images in `src` too.

I tried to run repo, but no luck, `env`/`modules`is missing, some deps are incompatibility.

You can try to set `$baseURL: '/public/images/';`

**Reply by JinHe-PerkinElmer** [reply]
At first, Thanks for your helping very much. The reason I remember doing this is to use the image path to access the image resource. For eg. our css in production like this: 
```
.arrow-btn-next {
    width: 32px;
    height: 32px;
    background: url(/images/icon_rightarrow.png) no-repeat;
    background-position: 50%;
}
<img alt=" " src="/images/header_icon_notification.svg">
```
We don't want to reference images by base64 code, which will increase the size of our package, since our package is already very large.

What is confusing is that the previous version(webpack v4) has been working fine, how can the new version(webpack v5) not work.

**Reply by alexander-akait** [reply]
> We don't want to reference images by base64 code, which will increase the size of our package, since our package is already very large.

You can configure so that this does not happen, just use `type: "asset/resource"`, not `type: "asset"`
